### PR TITLE
scripts: add a script to extract release notes

### LINKED ISSUE #20261: restore: cleanup is reliant on export storage availability
At present, when a restore resumes, it downloads or re-downloads the backup descriptor from the export storage. This is obviously necessary on the initial resume, and reasonable enough on subsequent resumes: if you're modifying your backup while doing a restore, you're just asking for trouble. 

If a restore is canceled or otherwise fails, we re-download the backup descriptor once more to determine what table descriptors to delete. This is less reasonable: we shouldn't be relying on the availability of the export storage to properly clean up after ourselves. Suppose, for example, someone runs a backup to S3, starts a restore, and then accidentally deletes that backup from S3. The restore will fail, but _also_ fail to clean up after itself because the backup descriptor will no longer be accessible. We should keep better track of what data we're writing so we can always clean it up, even if the backup descriptor goes away.

This is a harder project than it sounds. I was hopeful that simply knowing the table IDs would be sufficient because those are already stored in the jobs table payload. Sadly, because we simulate a drop table to get the schema changer to do our cleanup work for us, we need the entire table descriptor available. This would require storing the table descriptor in the jobs table, which we've so far avoided because of size concerns, as every progress update requires rewriting the entire jobs row.

I see two solutions:

   * Extract a separate `jobs_progress` table. We could use this opportunity to build support for "phased" jobs with separate progress bars that @danhhz has long dreamed of.

   * Change our restore cleanup to issue a `DelRange` on the affected spans, which should be derivable from the existing table rewrite information, instead of going through the schema changer.

### LINKED ISSUE #21236: ui: problemranges page never loads on local cluster
If I start a local cluster from scratch on current master (fa9a248617ffe4ab06f22c3b5cfef9165091a9a4) with fresh data directories and try to load the problemranges page, it hangs forever during the "Loading cluster status..." phase. There are no errors in the console.

I also saw this behavior when running on a 3-node cluster that had data in it when one node was down.

### LINKED ISSUE #21238: ui/debug: CommandQueue viz page should display errors
Errors on the `/reports/range/<rangeID>/cmdqueue` page are not displayed to the user. Instead, the page will continuously say "loading..." (see attached image). These errors can occur for a number of reasons, including that we're talking to a node that does not contain a replica in the range. We should display these errors prominently.

screen shot 2018-01-04 at 2 45 44 pm 2

### PR DESCRIPTION
**Pull request by knz**
Example use: 
```
$ python scripts/release-notes.py --from 64065087f53c59d1e23d32be6a8f2352b1abfa1c >example-notes.md
```

Resulting file `example-notes.md`:

## January 19, 2018

This release includes 78 merged PRs by 30 authors.
We would like to thank the following contributors from the CockroachDB community:

- Dmitry Saveliev
- Jincheng Li
- Yang Yuting
- 何羿宏

With special thanks to first-time contributors Jincheng Li, Yang Yuting, Paul Bardea, Abhishek Madan, Nathaniel Stewart and Andrew Kimball.

### General Changes

- added a mechanism to recompute range
  stats on demand. [#21345]
- CockroachDB now uses gRPC version 1.9.1 [#21398]

### Enterprise Changes

- make RESTORE cleanup resilient to
  external storage failure or removal
  
  Fixes #20261 [#21559]

### SQL Language Changes

- the output of EXPLAIN is enhanced when
  applied to statements containing scalar sub-queries. [#21305]
- Support ARRAY_AGG aggregation in distsql. [#21475]
- ON UPDATE CASCADE foreign key constraints are fully supported [#21329]
- Remove limit on transaction keys. There was
  formerly a limit of 100,000 keys. [#21078]
- improved error message when a column
  reference is ambiguous. [#21361]

### Admin UI Changes

- The command queue debug page now displays errors correctly.
  
  Fixes #21238. [#21529]
- Bug fix for the Problem Ranges debug page to again display
  all the problem ranges for the cluster.
  
  Fixes #21236. [#21522]

### Bug Fixes

- fixed incorrect columns and properties in
  EXPLAIN (VERBOSE) output. [#21527]
- fix a problem that could cause spurious GC
  activity, in particular after dropping a table. [#21407]
- Fix incorrect logic in lease rebalancing that
  prevented leases from being transferred [#21430]
- INSERT/UPDATE/DELETE ... RETURNING  ...
  statments used with the "statement data source" syntax (e.g. SELECT *
  FROM  [INSERT INTO ... RETURNING ...]) no longer ever commit the
  transaction, causing errors for the higher-level query. [#20847]

### Performance Improvements

- do less work during key decode [#21498]
- Speed up the performance of
  low-level delete operations. [#21507]
- small performance improvement to
  all scans that need to look at non-indexed columns. [#21459]
- Improve the performance of scans
  which encounter a large number of versions per key. [#21438]
- Improve low-level reverse scan
  throughput by up almost 4x when there are a few versions per key. [#21438]
- Queries on tables with many
  columns are more efficient. [#21450]
- Queries that only need to read
  part of a table with many columns are more efficient. [#21450]
- Improve low-level scan
  performance by 15-20% by disabling redundant checksums. [#21395]
- Reimplement low-level scan
  operations in C++, doubling performance for scans of contiguous
  keys/rows. [#21395]

### Build Changes

- The build system now enforces a minimum
  version of Go, rather than enforcing a specific version of Go. Since the
  Go 1.x series has strict backwards compatibility guarantees, the old
  rule was unnecessarily restrictive. [#21426]

#### Missing Category

- disallow non-transactional bulk IO statements inside transactions. [#21488]
- ALTER TABLE SCATTER now randomizes the locations of the
  leases for all ranges in the referenced table or index. [#21431]

#### Changes without release note annotation

- [#21547] [900291e48] sql: replace MakeJSON in Concat() and AsJSON() (何羿宏)
- [#21511] [187f8e662] opt: Separate data driven testing framework from tests (Andrew Kimball)
- [#21542] [9e25aeb8a] opt: Rename Expr to ExprView. (Andrew Kimball)
- [#20915] [4f33f0239] ui: remove spider from no results job view
 (Diana Hsieh)
- [#21513] [f8cb074e2] Update readme (Nathaniel Stewart)
- [#21513] [7b6e775b9] Update readme (Nathaniel Stewart)
- [#21513] [b74ce84bc] Update readme (Nathaniel Stewart)
- [#21132] [33d1f2749] sql: add json_agg (何羿宏)
- [#21478] [2d4883c12] opt: Integrate initial versions of v4 expr, memo, and factory (Andrew Kimball)
- [#21497] [d2e5fd351] libroach: add missing destructor (in parent class) (Raphael 'kena' Poss)
- [#21019] [1b136eb86] sql: support builtin json_build_object (Jincheng Li)
- [#21404] [20aa6386e] sqlccl: include http's configured custom CA setting when using custom s3 endpoints (David Taylor)
- [#21335] [cb1d819c5] sql: add json_strip_nulls (何羿宏)
- [#21457] [ec91158ed] opt: execution for selectOp (Radu Berinde)
- [#21395] [1d487c28f] storage/engine: reimplement MVCC{Scan,Get} in C++ (Peter Mattis)
- [#21395] [6193f3f9c] storage/engine: reimplement MVCC{Scan,Get} in C++ (Peter Mattis)
- [#21395] [563c01069] storage/engine: reimplement MVCC{Scan,Get} in C++ (Peter Mattis)

### PRs merged by committer

- Abhishek Madan:
  - [#21377] (+  71 -   8 ~  79/ 3) sql: Add overloads to ARRAY_AGG
  - [#21556] (+  33 -   0 ~  33/ 1) distsql: Implement diagramCellType for AlgebraicSetOpSpec
  - [#21475] (+  22 -   8 ~  30/ 5) distsql: Add ARRAY_AGG support

- Alex Robinson:
  - [#21430] (+ 102 -  21 ~ 123/ 3) storage: fix StoreList.filter to validate constraints properly
  - [#21431] (+ 171 - 121 ~ 292/ 6) *: Randomize lease placement in ALTER TABLE SCATTER (2 commits)

- Andrei Matei:
  - [#21532] (+   0 -   2 ~   2/ 1) sql: remove debug logging
  - [#21441] (+  76 -  56 ~ 132/ 6) sql: more directly init the timestamps in EvalCtx
  - [#21428] (+   1 -   0 ~   1/ 1) sql: make sure planner.autocommit gets reset
  - [#21486] (+ 242 - 119 ~ 361/22) sql: final assault on planner.session (6 commits)
  - [#20847] (+ 188 -  79 ~ 267/13) sql: don't allow StatementSources to commit the txn  (3 commits)
  - [#21408] (+ 951 - 744 ~1695/32) sql,pgwire: rework the implementation of the Copy-in pgwire subprotocol (2 commits)

- Andrew Couch:
  - [#21524] (+  10 -   6 ~  16/ 2) ui: ensure tooltips have spaces before all inline elements

- Andrew Kimball:
  - [#21511] (+ 265 - 221 ~ 486/ 2) opt: Separate data driven testing framework from tests
  - [#21542] (+ 451 - 449 ~ 900/ 6) opt: Rename Expr to ExprView.

- Andrew Kimball:
  - [#21478] (+7338 -   1 ~7339/17) opt: Integrate initial versions of v4 expr, memo, and factory

- Ben Darnell:
  - [#21484] (+   0 - 103 ~ 103/ 1) internal/client: Remove TestCommonMethods
  - [#21398] (+ 127 - 120 ~ 247/16) build: Update grpc to 1.9.1

- Bram Gruneir:
  - [#21529] (+ 133 - 104 ~ 237/ 8) ui: a number of small debug page fixes (5 commits)
  - [#21522] (+  17 -  14 ~  31/ 3) ui: Fixes the Problem Ranges Page to again display all node's problem ranges
  - [#21481] (+  39 -  74 ~ 113/ 9) sql: stop plumbing the bytes monitor down into the cascader
  - [#21329] (+2059 - 461 ~2520/18) sql: add support for ON UPDATE CASCADE for foreign key references
  - [#21554] (+  83 -   5 ~  88/ 3) sql/sqlbase: Fix the primary index span lookup for cascading

- David Taylor:
  - [#21404] (+  22 -   8 ~  30/ 1) sqlccl: include http's configured custom CA setting when using custom s3 endpoints
  - [#21488] (+ 106 -  34 ~ 140/ 7) sqlccl: disallow bulk IO statements in txn

- Diana Hsieh:
  - [#20915] (+  25 -  31 ~  56/ 3) ui: remove spider from no results job view

- Dmitry Saveliev:
  - [#21427] (+ 270 -  18 ~ 288/ 4) cli: add internal tables to expose status server info about the cluster

- Jordan Lewis:
  - [#21555] (+ 121 - 124 ~ 245/16) sqlbase: rename multirowfetcher to rowfetcher (2 commits)
  - [#21567] (+ 165 - 165 ~ 330/ 2) sqlbase: rename mrf to rf in rowfetcher
  - [#21450] (+ 101 -  47 ~ 148/ 3) sqlbase: multirowfetcher stops decoding a row when it doesn't need more data (4 commits)
  - [#21459] (+  15 -  10 ~  25/ 1) sqlbase: don't decode column family if unnecessary
  - [#21439] (+  70 -  17 ~  87/10) sqlbase: a few more optimizations on the fetch path (4 commits)
  - [#21422] (+  53 -  26 ~  79/ 5) sqlbase,encoding: a few small optimizations (4 commits)
  - [#21498] (+  44 -  18 ~  62/ 3) sql: don't decode the first key TableID/IndexID
  - [#21436] (+   4 -   4 ~   8/ 1) sql: deflake show_trace_replica logic test
  - [#21361] (+  32 -   8 ~  40/ 3) sql: better error message for ambiguous columns

- Justin Jaffray:
  - [#21019] (+ 128 -   0 ~ 128/ 3) sql: support builtin json_build_object (w/ Jincheng Li)
  - [#21547] (+  36 -   9 ~  45/ 3) sql: replace MakeJSON in Concat() and AsJSON() (w/ 何羿宏)
  - [#21519] (+ 102 -  38 ~ 140/ 2) sql: refactor check constraint validation
  - [#21335] (+ 296 -   0 ~ 296/ 6) sql: add json_strip_nulls (w/ 何羿宏)
  - [#21132] (+ 144 -   2 ~ 146/ 8) sql: add json_agg (w/ 何羿宏)

- Marc Berhault:
  - [#21469] (+   7 -   1 ~   8/ 2) libroach: add missing destructor.

- Matt Jibson:
  - [#21559] (+  15 -  16 ~  31/ 3) sqlccl: remove external storage dependency on RESTORE cleanup

- Matt Tracy:
  - [#21405] (+ 174 - 221 ~ 395/ 1) ts: Iterator Test Improvements.
  - [#21449] (+ 230 - 161 ~ 391/ 2) ts: Interpolation Max Distance (2 commits)
  - [#21491] (+ 144 -  30 ~ 174/ 4) ts: Interpolation Limit

- Nathan VanBenschoten:
  - [#21493] (+  39 -  11 ~  50/ 3) storage: handle lease inversion due to non-commutative lease equivalency
  - [#21533] (+  34 -   6 ~  40/ 2) retry: allow immediate retry from NextCh when isReset is true
  - [#21473] (+ 297 -  32 ~ 329/ 4) gossip: create SystemConfigDeltaFilter, use in SchemaChangeManager and LeaseManager (2 commits)

- Nathaniel Stewart:
  - [#21513] (+  16 -  83 ~  99/ 2) Update readme (3 commits)

- Nikhil Benesch:
  - [#21448] (+   7 -   1 ~   8/ 1) storage: suppress consistency logspam when quiescing
  - [#21392] (+  24 -  28 ~  52/ 1) sqlccl: speed up partitioning tests
  - [#21557] (+  27 -  25 ~  52/ 9) libroach,storage: make interval bounds on time-bound iterators sane
  - [#21426] (+  43 -  11 ~  54/ 3) build: only require a minimum version of Go

- Paul Bardea:
  - [#21520] (+   4 -   3 ~   7/ 1) docs: link to pprof wiki page from contributing.md
  - [#21479] (+ 361 -  26 ~ 387/ 7) distsql: add support for semijoins in HashJoiner

- Pete Vilter:
  - [#21517] (+   6 -   4 ~  10/ 1) ui: tweak capacity tooltip (spaces were missing)
  - [#21568] (+  13 -   0 ~  13/ 2) ui: teach event list about sequence event types
  - [#21541] (+ 134 -   6 ~ 140/ 6) sql: sequences: add setval(name, value, is_called) builtin
  - [#21540] (+ 153 -   8 ~ 161/ 9) sql: allow SELECTing from a sequence

- Peter Mattis:
  - [#21395] (+1108 - 119 ~1227/11) storage/engine: reimplement MVCC{Scan,Get} in C++ (5 commits)
  - [#21507] (+  61 -  25 ~  86/ 2) storage/engine: reimplement MVCCDeleteRange in terms of MVCCScan
  - [#21438] (+ 185 -  58 ~ 243/ 2) libroach: optimize MVCC reverse scans (3 commits)

- Radu Berinde:
  - [#21512] (+ 354 -   0 ~ 354/ 2) util: data structure for small int-to-int mappings
  - [#21527] (+1810 -1810 ~3620/29) sql: fix EXPLAIN (VERBOSE) output
  - [#21457] (+ 400 - 220 ~ 620/16) opt: execution for selectOp (5 commits)

- Raphael 'kena' Poss:
  - [#21118] (+  76 -  49 ~ 125/ 1) sql: reduce allocations in show_create.go (w/ Yang Yuting)
  - [#21472] (+   1 -   1 ~   2/ 1) sql: minor fix to show_create.go
  - [#21305] (+ 741 - 623 ~1364/30) sql: improve the handling of subqueries (10 commits)
  - [#21497] (+  22 -  16 ~  38/ 2) libroach: add missing destructor (in parent class) (3 commits)

- Rebecca Taft:
  - [#21352] (+ 525 - 129 ~ 654/11) opt: Add support for building filters in the `Expr` tree

- Spencer Kimball:
  - [#21078] (+ 930 - 344 ~1274/36) storage, kv: remove the max intent limit

- Tobias Schottdorf:
  - [#21345] (+ 711 - 135 ~ 846/23) storage: recompute and adjust stats via consistency checker (2 commits)
  - [#21407] (+ 980 - 417 ~1397/ 4) storage: fix (the semantics of) MVCCStats.GCBytesAge
  - [#21466] (+  57 -   0 ~  57/ 6) storage: don't panic on unknown (batch) request

- Vivek Menezes:
  - [#21550] (+  17 -   7 ~  24/ 2) util/mon: improve memory budget exceeded error message
  - [#21411] (+   9 -   0 ~   9/ 1) ui: add graphs for clock offset mean and stddev
  - [#21418] (+   3 -   2 ~   5/ 1) sql: fix SCRUB double call to RowContainer.Close()
  - [#21546] (+  27 -  30 ~  57/ 3) sql: do not pass testing.T to RunScrub()

[#20847]: https://github.com/cockroachdb/cockroach/pull/20847
[#20915]: https://github.com/cockroachdb/cockroach/pull/20915
[#21019]: https://github.com/cockroachdb/cockroach/pull/21019
[#21078]: https://github.com/cockroachdb/cockroach/pull/21078
[#21118]: https://github.com/cockroachdb/cockroach/pull/21118
[#21132]: https://github.com/cockroachdb/cockroach/pull/21132
[#21305]: https://github.com/cockroachdb/cockroach/pull/21305
[#21329]: https://github.com/cockroachdb/cockroach/pull/21329
[#21335]: https://github.com/cockroachdb/cockroach/pull/21335
[#21345]: https://github.com/cockroachdb/cockroach/pull/21345
[#21352]: https://github.com/cockroachdb/cockroach/pull/21352
[#21361]: https://github.com/cockroachdb/cockroach/pull/21361
[#21377]: https://github.com/cockroachdb/cockroach/pull/21377
[#21392]: https://github.com/cockroachdb/cockroach/pull/21392
[#21395]: https://github.com/cockroachdb/cockroach/pull/21395
[#21398]: https://github.com/cockroachdb/cockroach/pull/21398
[#21404]: https://github.com/cockroachdb/cockroach/pull/21404
[#21405]: https://github.com/cockroachdb/cockroach/pull/21405
[#21407]: https://github.com/cockroachdb/cockroach/pull/21407
[#21408]: https://github.com/cockroachdb/cockroach/pull/21408
[#21411]: https://github.com/cockroachdb/cockroach/pull/21411
[#21418]: https://github.com/cockroachdb/cockroach/pull/21418
[#21422]: https://github.com/cockroachdb/cockroach/pull/21422
[#21426]: https://github.com/cockroachdb/cockroach/pull/21426
[#21427]: https://github.com/cockroachdb/cockroach/pull/21427
[#21428]: https://github.com/cockroachdb/cockroach/pull/21428
[#21430]: https://github.com/cockroachdb/cockroach/pull/21430
[#21431]: https://github.com/cockroachdb/cockroach/pull/21431
[#21436]: https://github.com/cockroachdb/cockroach/pull/21436
[#21438]: https://github.com/cockroachdb/cockroach/pull/21438
[#21439]: https://github.com/cockroachdb/cockroach/pull/21439
[#21441]: https://github.com/cockroachdb/cockroach/pull/21441
[#21448]: https://github.com/cockroachdb/cockroach/pull/21448
[#21449]: https://github.com/cockroachdb/cockroach/pull/21449
[#21450]: https://github.com/cockroachdb/cockroach/pull/21450
[#21457]: https://github.com/cockroachdb/cockroach/pull/21457
[#21459]: https://github.com/cockroachdb/cockroach/pull/21459
[#21466]: https://github.com/cockroachdb/cockroach/pull/21466
[#21469]: https://github.com/cockroachdb/cockroach/pull/21469
[#21472]: https://github.com/cockroachdb/cockroach/pull/21472
[#21473]: https://github.com/cockroachdb/cockroach/pull/21473
[#21475]: https://github.com/cockroachdb/cockroach/pull/21475
[#21478]: https://github.com/cockroachdb/cockroach/pull/21478
[#21479]: https://github.com/cockroachdb/cockroach/pull/21479
[#21481]: https://github.com/cockroachdb/cockroach/pull/21481
[#21484]: https://github.com/cockroachdb/cockroach/pull/21484
[#21486]: https://github.com/cockroachdb/cockroach/pull/21486
[#21488]: https://github.com/cockroachdb/cockroach/pull/21488
[#21491]: https://github.com/cockroachdb/cockroach/pull/21491
[#21493]: https://github.com/cockroachdb/cockroach/pull/21493
[#21497]: https://github.com/cockroachdb/cockroach/pull/21497
[#21498]: https://github.com/cockroachdb/cockroach/pull/21498
[#21507]: https://github.com/cockroachdb/cockroach/pull/21507
[#21511]: https://github.com/cockroachdb/cockroach/pull/21511
[#21512]: https://github.com/cockroachdb/cockroach/pull/21512
[#21513]: https://github.com/cockroachdb/cockroach/pull/21513
[#21517]: https://github.com/cockroachdb/cockroach/pull/21517
[#21519]: https://github.com/cockroachdb/cockroach/pull/21519
[#21520]: https://github.com/cockroachdb/cockroach/pull/21520
[#21522]: https://github.com/cockroachdb/cockroach/pull/21522
[#21524]: https://github.com/cockroachdb/cockroach/pull/21524
[#21527]: https://github.com/cockroachdb/cockroach/pull/21527
[#21529]: https://github.com/cockroachdb/cockroach/pull/21529
[#21532]: https://github.com/cockroachdb/cockroach/pull/21532
[#21533]: https://github.com/cockroachdb/cockroach/pull/21533
[#21540]: https://github.com/cockroachdb/cockroach/pull/21540
[#21541]: https://github.com/cockroachdb/cockroach/pull/21541
[#21542]: https://github.com/cockroachdb/cockroach/pull/21542
[#21546]: https://github.com/cockroachdb/cockroach/pull/21546
[#21547]: https://github.com/cockroachdb/cockroach/pull/21547
[#21550]: https://github.com/cockroachdb/cockroach/pull/21550
[#21554]: https://github.com/cockroachdb/cockroach/pull/21554
[#21555]: https://github.com/cockroachdb/cockroach/pull/21555
[#21556]: https://github.com/cockroachdb/cockroach/pull/21556
[#21557]: https://github.com/cockroachdb/cockroach/pull/21557
[#21559]: https://github.com/cockroachdb/cockroach/pull/21559
[#21567]: https://github.com/cockroachdb/cockroach/pull/21567
[#21568]: https://github.com/cockroachdb/cockroach/pull/21568
[b74ce84bc]: https://github.com/cockroachdb/cockroach/commit/b74ce84bc
[ec91158ed]: https://github.com/cockroachdb/cockroach/commit/ec91158ed
[900291e48]: https://github.com/cockroachdb/cockroach/commit/900291e48
[cb1d819c5]: https://github.com/cockroachdb/cockroach/commit/cb1d819c5
[1b136eb86]: https://github.com/cockroachdb/cockroach/commit/1b136eb86
[6193f3f9c]: https://github.com/cockroachdb/cockroach/commit/6193f3f9c
[187f8e662]: https://github.com/cockroachdb/cockroach/commit/187f8e662
[33d1f2749]: https://github.com/cockroachdb/cockroach/commit/33d1f2749
[563c01069]: https://github.com/cockroachdb/cockroach/commit/563c01069
[2d4883c12]: https://github.com/cockroachdb/cockroach/commit/2d4883c12
[9e25aeb8a]: https://github.com/cockroachdb/cockroach/commit/9e25aeb8a
[7b6e775b9]: https://github.com/cockroachdb/cockroach/commit/7b6e775b9
[20aa6386e]: https://github.com/cockroachdb/cockroach/commit/20aa6386e
[1d487c28f]: https://github.com/cockroachdb/cockroach/commit/1d487c28f
[f8cb074e2]: https://github.com/cockroachdb/cockroach/commit/f8cb074e2
[4f33f0239]: https://github.com/cockroachdb/cockroach/commit/4f33f0239
[d2e5fd351]: https://github.com/cockroachdb/cockroach/commit/d2e5fd351

### CODE REVIEW

**Comment by knz** [comment]
I welcome suggestions to alter the output in a way that's most useful to you.

I can also investigate how to print out the PR number next to the release note based on the advice Ben gave earlier.

**Comment by knz** [comment]
Also I can have the release notes grouped by category if you think that's useful.

**Comment by benesch** [comment]
Code :lgtm:

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[scripts/release-notes.sh, line 5 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-L345Fdzv8d1Cr7k5EBp:-L345Fe-02jhOgQS6ijb:b2nhoia) ([raw file](https://github.com/cockroachdb/cockroach/blob/4359601aefc56defc46ad42e45bb4ff97e375b67/scripts/release-notes.sh#L5)):*
> ```Shell
> set -euo pipefail
> 
> range=${1?specify range, e.g. v2.0-alpha.20171218...1b2eabdd856263662cef7f0850ad74ef669cb2d9}
> ```

TIL

---

*[scripts/release-notes.sh, line 7 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-L345JaDLevyuzAQ0rvv:-L345JaDLevyuzAQ0rvw:b-al0z8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/4359601aefc56defc46ad42e45bb4ff97e375b67/scripts/release-notes.sh#L7)):*
> ```Shell
> range=${1?specify range, e.g. v2.0-alpha.20171218...1b2eabdd856263662cef7f0850ad74ef669cb2d9}
> 
> git log $range --pretty='%H %s' --no-merges | (
> ```

Mind keeping shellcheck happy and double-quoting `$range` here? (And `$commit` a few lines below.)

Optional also: no need to spawn a subshell; you can use `{` and `}` grouping instead.

---

*[scripts/release-notes.sh, line 16 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-L34625j8lu_VVN07Ovt:-L34625kYExcoWlLyvs4:be9xraq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4359601aefc56defc46ad42e45bb4ff97e375b67/scripts/release-notes.sh#L16)):*
> ```Shell
> 	    continue
> 	fi
> 	notes=$(echo "$message" | grep -A 10000 -E '^[rR]elease note[^:]*:' || true)
> ```

optional: I find `sed` cleaner here:

```
sed -nE '/^[rR]elease note/,$p'
```

---

*[scripts/release-notes.sh, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-L346_2b8MiGCbZ4eB3N:-L346_2b8MiGCbZ4eB3O:benhw2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/4359601aefc56defc46ad42e45bb4ff97e375b67/scripts/release-notes.sh#L19)):*
> ```Shell
> 	if test -z "$notes"; then
> 	    # Missing 'release note' line. Keep track for later.
> 	    nonotes=$(echo "$nonotes"; echo "$commit $title")
> ```

Optional alternative:

```
nonotes+="$commit $title"
nonotes+=$'\n'
```

---

*[scripts/release-notes.sh, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-L346GbzMpys_Fpfrz6Q:-L346GbzMpys_Fpfrz6R:b-79ksu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4359601aefc56defc46ad42e45bb4ff97e375b67/scripts/release-notes.sh#L25)):*
> ```Shell
> 	echo "$notes" | sed -e 's/^/    /g'
> 	echo
> 	counter=$(expr $counter + 1)
> ```

Optional alternative:

```
((counter++))
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-:-L3494azP4cMT-wtYUvi:b-wyr02u)*

**Comment by knz** [comment]
I have rewritten the program entirely to:

1) categorize the release notes automatically
2) fix misspells in the release note categories
3) print out the release notes document in valid markdown, so that only minimal editing is left to be done
4) append the per-contributor list which I wanted to have for a long time

**Comment by knz** [comment]
See an example output in the new PR description at the top.

**Comment by jseldess** [comment]
This is amazing, @knz! Thank you. I'll test it soon and get back with feedback/ideas.

**Comment by jseldess** [comment]
Just tested this, and it works wonderfully. I'll let other devs review the code, but a few requests/questions:

- Historically, we've only provide the PR link per change, not the commit SHA. I'd prefer to stick to that approach. 
- I like adding the list of external contributors back into release notes, but I think listing cockroach employees is unnecessary and listing each contributor's commits is overkill. I'd prefer just a bulleted list of external contributors with an intro paragraph summarizing `## merged PRs by ## authors`. It would be nice to thank first-time contributors, if we're able. For example:

> This release includes 89 merged PRs by 19 authors. We would like to thank the following contributors from the CockroachDB community, especially first-time contributor Karl Southern.
>
> - Karl Southern
> - Kenjiro Nakayama

Thoughts?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-:-L39llpBd3AoZBANZ3hh:bbrco31)*

**Comment by jseldess** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[scripts/release-notes.py, line 5 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-L39lIRVv_mEUsAWUaUC:-L39lIRVv_mEUsAWUaUD:b-eias48) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ce804e32b94153dd2218e251e9678b88af742cb/scripts/release-notes.py#L5)):*
> ```Python
> # Show a compact ChangeLog representation of a range of Git commits.
> #
> # Example use: changelog.py -f v1.1-alpha.20170817
> ```

This should be `release-notes.py`, not `changelog.py`. Also, it'd be more helpful to show both `-f` and `-t ` since we always are looking for commits `-f` a specific tag `-t` the candidate SHA.

---

*[scripts/release-notes.py, line 41 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-L39kxdBKRW7XeGqheS5:-L39kxdBKRW7XeGqheS6:b66mzck) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ce804e32b94153dd2218e251e9678b88af742cb/scripts/release-notes.py#L41)):*
> ```Python
>     'admin ui change': "Admin UI Changes",
>     'general change': "General Changes",
>     'core change': "Core Changes",
> ```

I don't think we've ever had a "Core Changes" section in release notes. These tend to get put in "General Changes".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515)*

**Comment by knz** [comment]
1) you can disable the commit SHA using the new option `--hide-unambiguous-shas`.
2) I have added the summary line "Merged N PRs by N Authors" at the top, together with "We'd like to thank the following contributors from the CockroachDB community" and the list of external authors.
3) This program structure cannot detect first-time contributors. We'd have to create a database and maintain it over time. Perhaps someone can pick this up.
4) You can cut out the per-contributor list away from the generated document. It's easier to take stuff out than to bake it in.

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-:-L3Ecekg27BmyZHp9KN9:bc9ctv2)*

**Comment by knz** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[scripts/release-notes.py, line 5 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-L39lIRVv_mEUsAWUaUC:-L3Emq8NPMQAtrQc414-:b-xo95ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ce804e32b94153dd2218e251e9678b88af742cb/scripts/release-notes.py#L5)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

This should be `release-notes.py`, not `changelog.py`. Also, it'd be more helpful to show both `-f` and `-t ` since we always are looking for commits `-f` a specific tag `-t` the candidate SHA.
</blockquote></details>

Pointed to `--help` instead.

---

*[scripts/release-notes.py, line 41 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-L39kxdBKRW7XeGqheS5:-L3EmhS8yaoEd8AGNA0M:bwn5jdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ce804e32b94153dd2218e251e9678b88af742cb/scripts/release-notes.py#L41)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

I don't think we've ever had a "Core Changes" section in release notes. These tend to get put in "General Changes".
</blockquote></details>

Some folk have used "`Release note (core change): ...`" but it's ok, I folded it into general changes. see below

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515)*

**Comment by jseldess** [comment]
All makes sense, Raphael. Thank you so much! :lgtm:

One side-note, when you leave SHAs in, I don't think the reference-style links work. Without the SHAs, works perfectly, though.

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21515#-:-L3EqcVlZgxFoHD9V6l2:bqr7aj7)*

**Comment by knz** [comment]
There was a problem in the generated markdown. Good that you caught it. Fixed.

**Comment by bdarnell** [comment]
LGTM

> This program structure cannot detect first-time contributors. We'd have to create a database and maintain it over time.

Git is that database. When I was doing the release notes and included this section, I would just do `git log --author` for each external contributor and see if there was anything before the current release. Automating it would take a little work, but most releases have few enough external contributors that it's reasonable to do it by hand if this is a detail we want to include.

Now that this script exists, I think it would be appropriate for the commit message linter to get more strict about the format of the release note.

**Comment by knz** [comment]
I have modified the script to also list first-time contributors using the tip given by Ben.
Note that I have chosen to also include CRL members in the list of first-time contributors, if they have indeed contributed for the first time in the given period. That seems fair.

The release note format is now linted more strictly, see #21588.

**Comment by knz** [comment]
@jseldess I forgot to mention the `--from` commit is *excluded* from the analysis. This makes it possible to use e.g. `--from <prev release tag> --until <candidate new release sha>`.

**Comment by jseldess** [comment]
Perfect. Thanks you, @knz. This will save me (and whoever else ends up doing release notes) a lot of time.

File: scripts/release-notes.py [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+
+ext_contributors = individual_authors - crdb_folk
+if len(ext_contributors) > 0:
+    ext_contributors = sorted(ext_contributors)
+    print("We would like to thank the following contributors from the CockroachDB community:")
+    print()
+    for a in ext_contributors:
+        print("-", a)
+    print()
+# Note: CRDB folk can be first-time contributors too, so
```
Comment (bdarnell) [inline scripts/release-notes.py]: It's confusing to see first-time contributors who don't appear in the preceding list (as in [this week's release)](https://www.cockroachlabs.com/docs/releases/v2.0-alpha.20180129.html)). I think we should either exclude CRDB employees from the first-timers list (my first choice) or say something about who these first-time non-community contributors are.

Comment (knz) [reply scripts/release-notes.py]: Following up in #22206.

**Pull request merged**